opentracing/recorder.go

```diff
@@ -0,0 +1,90 @@
+package opentracing
+
+import (
+	"fmt"
+	"log"
+	"sync"
+
+	basictracer "github.com/opentracing/basictracer-go"
+	"sourcegraph.com/sourcegraph/appdash"
+)
+
+// Recorder implements basictracer.Recorder.
+type Recorder struct {
+	collector appdash.Collector
+	logOnce   sync.Once
+	verbose   bool
+}
+
+// NewRecorder forwards basictracer.RawSpans to an appdash.Collector.
+func NewRecorder(collector appdash.Collector, opts Options) *Recorder {
+	return &Recorder{
+		collector: collector,
+		verbose:   opts.Verbose,
+	}
+}
+
+// RecordSpan converts a RawSpan into the Appdash representation of a span
+// and records it to the underlying collector.
+func (r *Recorder) RecordSpan(sp basictracer.RawSpan) {
+	spanID := appdash.SpanID{
+		Span:   appdash.ID(uint64(sp.SpanID)),
+		Trace:  appdash.ID(uint64(sp.TraceID)),
+		Parent: appdash.ID(uint64(sp.ParentSpanID)),
+	}
+
+	r.collectEvent(spanID, appdash.SpanName(sp.Operation))
+
+	// Record all of the logs. Payloads are thrown out.
+	for _, log := range sp.Logs {
+		r.collectEvent(spanID, appdash.LogWithTimestamp(log.Event, log.Timestamp))
+	}
+
+	for key, value := range sp.Tags {
+		val := []byte(fmt.Sprintf("%v", value))
+		r.collectAnnotation(spanID, appdash.Annotation{Key: key, Value: val})
+	}
+
+	for key, val := range sp.Baggage {
+		r.collectAnnotation(spanID, appdash.Annotation{Key: key, Value: []byte(val)})
```

Comment:
> I wonder if it's left up to the user to attach these as metadata with e.g. gRPC requests or as headers on HTTP requests, or if they do something else?

Yeah it's the user's responsibility to propagate a Span's metadata, which in the case of Appdash is the TraceID and SpanID, and its baggage downstream. If you look at [this line in the example](https://github.com/sourcegraph/appdash/pull/110/files#diff-957e1ccc381c6518748cd1a86af2d2d5R50), this is done by "injecting" the Span into the request headers. On [the server side](https://github.com/sourcegraph/appdash/pull/110/files#diff-957e1ccc381c6518748cd1a86af2d2d5R64), the metadata and baggage is extracted, and then a new span is created and joins the trace using that data.